---
title: "R Notebook"
output: html_notebook
---

## Load libraries and scripts
```{r}

if(!require(pacman)) {
  install.packages("pacman")
  library(pacman)
}

p_load(tidyverse)
p_load(GO.db)
p_load(iheatmapr)		
p_load(viridis)
p_load(readxl)

source( '../Common/global_parameters.R')
source( '../Common/helper_functions.R')

sessionInfo()

```

## Directories Management
```{r}
soms_results_dir <- file.path(results_dir, "Self_Organising_Maps/EdgeR_RUVs_QL/To_Use_k_0_Remove_5a/Condition_on_child_terms_true" ) 

```

## Read list of enriched GO terms related to each SOM cluster
```{r}

## Enriched GO terms per cluster
soms_A_go_results <- read_tsv( file.path(soms_results_dir, "compare_A/5x5/enriched_go_terms_SOMs.txt"))

soms_AL_go_results <- read_tsv( file.path(soms_results_dir, "compare_AL/5x5/enriched_go_terms_SOMs.txt"))

## Genes associated with enriched GO terms per cluster
soms_A_go_genes <- read_tsv( file.path(soms_results_dir, "compare_A/5x5/enriched_genes_SOMs_updated.txt")) %>%
                   mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   ) %>%
                   mutate( cluster_id = as.integer(cluster_id))

soms_AL_go_genes <- read_tsv( file.path(soms_results_dir, "compare_AL/5x5/enriched_genes_SOMs_updated.txt")) %>%
                    mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   )  %>%
                    mutate( cluster_id = as.integer(cluster_id))

## Direction of fold-change per cluster
soms_A_go_results_cleaned <- soms_A_go_results %>%
  filter( pvalue < 0.01 & ontology == "BP") %>%
  mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   ) %>%
  mutate( cluster_id = as.integer(cluster_id)) %>%
  dplyr::select( gobpid, term, cluster_id, pvalue, bh  ) %>%
  mutate( Enrichment = case_when( bh < 0.05 ~ "High",
                                  pvalue < 0.01 & bh >= 0.05 ~ "Medium",
                                  TRUE ~ "Low")) %>%
  mutate( Enrichment_Levels = case_when( bh < 0.05 ~ 2,
                                  pvalue < 0.01 & bh >= 0.05 ~ 1,
                                  TRUE ~ 0)) %>%  
  distinct()


soms_AL_go_results_cleaned <- soms_AL_go_results %>%
  filter( pvalue < 0.01 & ontology == "BP") %>%
  mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   ) %>%
  mutate( cluster_id = as.integer(cluster_id)) %>%
  dplyr::select( gobpid, term, cluster_id, pvalue, bh  ) %>%
  mutate( Enrichment = case_when( bh < 0.05 ~ "High",
                                  pvalue < 0.01 & bh >= 0.05 ~ "Medium",
                                  TRUE ~ "Low")) %>%
  mutate( Enrichment_Levels = case_when( bh < 0.05 ~ 2,
                                  pvalue < 0.01 & bh >= 0.05 ~ 1,
                                  TRUE ~ 0)) %>%  
  distinct()


```

## Gene per SOM cluster and the standardised gene expression value for each sample
Merge the gene per cluster information with the gene annotation information
```{r}
compare_A_soms <- read.table(file.path(soms_results_dir, "compare_A/5x5/cluster_for_each_gene.tab"), stringsAsFactors = FALSE )

compare_AL_soms <- read.table(file.path(soms_results_dir, "compare_AL/5x5/cluster_for_each_gene.tab"), stringsAsFactors = FALSE)

gene_annotation <- read_tsv( file.path(data_dir, "Gene_Annotation/cneo_gene_annotation_summary.txt" ))


compare_A_soms_annot <- compare_A_soms %>%
  dplyr::distinct( id_a, cluster) %>%
  left_join( gene_annotation, by = c( "id_a" = "orf_id")) %>%
  dplyr::distinct( id_a, cluster, is_tf, is_kinases, is_virulence_gene, order_peaktime, scer_orthologs)


compare_AL_soms_annot <- compare_AL_soms %>%
  dplyr::distinct( id_a, cluster) %>%
  left_join( gene_annotation, by = c( "id_a" = "orf_id")) %>%
  dplyr::distinct( id_a, cluster, is_tf, is_kinases, is_virulence_gene, order_peaktime, scer_orthologs)


```


## Summarise gene annotation per cluster 
```{r}

compare_A_col_annot <- compare_A_soms_annot  %>%
                       group_by (cluster  ) %>%
                        summarise( count_tfs =   sum(ifelse(is.na(is_tf), 0, 1)),
                                   count_kinases =   sum(ifelse(is.na(is_kinases), 0, 1)),
                                   count_virulence_genes = sum(ifelse(is.na(is_virulence_gene), 0, 1)),
                                   count_cell_cycle_genes = sum(ifelse(is.na(order_peaktime), 0, 1))) %>%
                        filter(  count_tfs > 0 |
                                  count_kinases > 0 |
                                  count_virulence_genes > 0 |
                                  count_cell_cycle_genes > 0  )


compare_AL_col_annot <- compare_AL_soms_annot  %>%
                       group_by (cluster  ) %>%
                        summarise( count_tfs =   sum(ifelse(is.na(is_tf), 0, 1)),
                                   count_kinases =   sum(ifelse(is.na(is_kinases), 0, 1)),
                                   count_virulence_genes = sum(ifelse(is.na(is_virulence_gene), 0, 1)),
                                   count_cell_cycle_genes = sum(ifelse(is.na(order_peaktime), 0, 1))) %>%
                        filter(  count_tfs > 0 |
                                  count_kinases > 0 |
                                  count_virulence_genes > 0 |
                                  count_cell_cycle_genes > 0  )

```



## Read list of selected GO terms
```{r}
selected_GO_terms <- readxl::read_xlsx( file.path(data_dir, "Data_for_SOMs/Selected SOMs enriched terms.xlsx" ), sheet="Lookup_Table" ) 
  
```
## Read in the table showing direction of gene expression (for the genes in the cluster)
```{r}
compare_A_direction <- read_tsv ( file.path( soms_results_dir,"compare_A/5x5/soms_clusters_mapping.txt"  )) %>%
                       mutate( Direction = case_when( str_detect(Direction, "Up")  ~ "Up",
                                                      str_detect(Direction, "Down")  ~ "Down",
                                                      TRUE ~ NA_character_
                                                      ) )        
  
compare_AL_direction <- read_tsv ( file.path( soms_results_dir,"compare_AL/5x5/soms_clusters_mapping.txt"  )) %>%
                       mutate( Direction = case_when( str_detect(Direction, "Up") ~ "Up",
                                                      str_detect(Direction, "Down") ~ "Down",
                                                      TRUE ~ NA_character_
                                                      ) )        

```

```{r}

soms_A_select_go <- soms_A_go_results_cleaned %>%
    inner_join( selected_GO_terms, by = c( "gobpid" = "GO_ID") ) %>%
    filter( Treatment == "AMB treatment" ) %>%
    arrange( Responses, GO_term  ) %>%
    dplyr::select( - Cluster) %>%
    distinct()

soms_A_select_go_matrix <- soms_A_select_go %>% 
    dplyr::select( Responses, cluster_id, term, Enrichment_Levels ) %>%
     arrange( cluster_id, Responses, term  ) %>%
    distinct() %>%
    spread( cluster_id, Enrichment_Levels, fill = NA)  %>%
    arrange( Responses, term  ) %>%
    dplyr::select( - Responses) %>%
    distinct( ) %>%
    as.data.frame %>%
    column_to_rownames("term") %>%
    as.matrix()

soms_A_group_term <- soms_A_select_go %>%
      distinct( Responses, GO_term  ) %>%
      arrange( Responses, GO_term  ) 


soms_A_direction <-  soms_A_select_go %>%
                      distinct( cluster_id) %>%
                      inner_join ( compare_A_direction, by=c("cluster_id" = "cluster" ) ) %>%
                      distinct( cluster_id, Direction) %>%
                      arrange( cluster_id)
                     
if( any ( colnames( soms_A_select_go_matrix) != soms_A_direction$cluster_id )) {
  stop("Order of SOMs cluster not matched for merging with table describing the direction of gene expression change.")
}

cneo_soms_A_select <- main_heatmap(  soms_A_select_go_matrix, colors= c("lightblue", "darkblue"), layout = list(margin = list(l = 500)) )  %>%
  add_row_labels() %>%
  add_col_labels() %>%
  add_row_annotation( data.frame( Responses =  soms_A_group_term$Responses)) %>%
  add_col_annotation( data.frame( Direction = soms_A_direction$Direction), colors = list(Direction=c("green", "red")))

cneo_soms_A_select

dim(soms_A_select_go_matrix)

save_iheatmap(p=cneo_soms_A_select, filename=file.path( soms_results_dir, "compare_A/5x5/compare_A_selected_go_terms_iheatmapr.png"))

```


```{r}

soms_A_gene_annotation <- soms_A_select_go %>%
  inner_join  ( soms_A_go_genes, by =c( "cluster_id" = "cluster_id",
                                       "gobpid" = "go_id") ) %>%
  dplyr::distinct( gobpid, term, cluster_id, Treatment, gene_id, is_tf, is_kinases, is_virulence_gene, order_peaktime)

soms_A_gene_row_annot_sum <- soms_A_gene_annotation %>%
  dplyr::select( -cluster_id, -Treatment) %>%
  dplyr::distinct( ) %>%
  group_by( term ) %>%
  summarise( count_tfs =   sum(ifelse(is.na(is_tf), 0, 1)),
             count_kinases =   sum(ifelse(is.na(is_kinases), 0, 1)),
             count_virulence_genes = sum(ifelse(is.na(is_virulence_gene), 0, 1)),
             count_cell_cycle_genes = sum(ifelse(is.na(order_peaktime), 0, 1)))  %>%
  # filter(  count_tfs > 0 |
  #           count_kinases > 0 |
  #           count_virulence_genes > 0 |
  #           count_cell_cycle_genes > 0 ) %>%
 arrange(term) %>%
 column_to_rownames( "term")

dim(soms_A_select_go_matrix)
dim(soms_A_gene_row_annot_sum )

soms_A_gene_col_annot_sum <- compare_A_col_annot %>%
                             filter( cluster %in% (soms_A_gene_annotation %>% pull(cluster_id) %>% unique) ) %>%
                             column_to_rownames("cluster")
                         

cneo_soms_A_select_annot <- cneo_soms_A_select %>%
                            add_row_annotation(  soms_A_gene_row_annot_sum) # %>%
                            # add_col_annotation( soms_A_gene_col_annot_sum)

cneo_soms_A_select_annot
```



## AMB + LF treatment
```{r}

soms_AL_select_go <- soms_AL_go_results_cleaned %>%
    inner_join( selected_GO_terms, by = c( "gobpid" = "GO_ID") ) %>%
    filter( Treatment == "AMB+LF treatment" ) %>%
    arrange( Responses, GO_term  ) %>%
    dplyr::select( - Cluster) %>%
    distinct()

soms_AL_select_go_matrix <- soms_AL_select_go %>% 
    dplyr::select( Responses, cluster_id, term, Enrichment_Levels ) %>%
     arrange( cluster_id, Responses, term  ) %>%
    distinct() %>%
    spread( cluster_id, Enrichment_Levels, fill = NA)  %>%
    arrange( Responses, term  ) %>%
    dplyr::select( - Responses) %>%
    distinct( ) %>%
    as.data.frame %>%
    column_to_rownames("term") %>%
    as.matrix()

soms_AL_group_term <- soms_AL_select_go %>%
      distinct( Responses, GO_term  ) %>%
      arrange( Responses, GO_term  ) 


soms_AL_direction <-  soms_AL_select_go %>%
                      distinct( cluster_id) %>%
                      inner_join ( compare_AL_direction, by=c("cluster_id" = "cluster" ) ) %>%
                      distinct( cluster_id, Direction) %>%
                      arrange( cluster_id)
                     
if( any ( colnames( soms_AL_select_go_matrix) != soms_AL_direction$cluster_id )) {
  stop("Order of SOMs cluster not matched for merging with table describing the direction of gene expression change.")
}

cneo_soms_AL_select <- main_heatmap(  soms_AL_select_go_matrix, colors= c("lightblue", "darkblue"), layout = list(margin = list(l = 500)) )  %>%
  add_row_labels() %>%
  add_col_labels() %>%
  add_row_annotation( data.frame( Responses =  soms_AL_group_term$Responses)) %>%
  add_col_annotation( data.frame( Direction = soms_AL_direction$Direction), colors = list(Direction=c("green", "red")))

cneo_soms_AL_select

save_iheatmap(p=cneo_soms_AL_select, filename=file.path( soms_results_dir, "compare_AL/5x5/compare_AL_selected_go_terms_iheatmapr.png"))

dim(soms_AL_select_go_matrix)
```




```{r}

soms_AL_gene_annotation <- soms_AL_select_go %>%
  inner_join  ( soms_AL_go_genes, by =c( "cluster_id" = "cluster_id",
                                       "gobpid" = "go_id") ) %>%
  dplyr::distinct( gobpid, term, cluster_id, Treatment, gene_id, is_tf, is_kinases, is_virulence_gene, order_peaktime)

soms_AL_gene_row_annot_sum <- soms_AL_gene_annotation %>%
  dplyr::select( -cluster_id, -Treatment ) %>%
  dplyr::distinct( ) %>%
  group_by( term ) %>%
  summarise(  count_tfs =   sum(ifelse(is.na(is_tf), 0, 1)),
             count_kinases =   sum(ifelse(is.na(is_kinases), 0, 1)),
             count_virulence_genes = sum(ifelse(is.na(is_virulence_gene), 0, 1)),
             count_cell_cycle_genes = sum(ifelse(is.na(order_peaktime), 0, 1))) %>%
  # filter(  count_tfs > 0 |
  #           count_kinases > 0 |
  #           count_virulence_genes > 0 |
  #           count_cell_cycle_genes > 0 ) %>%
 arrange(term) %>%
 column_to_rownames( "term")

soms_AL_gene_col_annot_sum <- compare_AL_col_annot %>%
                             filter( cluster %in% (soms_AL_gene_annotation %>% pull(cluster_id) %>% unique) ) %>%
                             column_to_rownames("cluster")

cneo_soms_AL_select_annot <- cneo_soms_AL_select %>%
                            add_row_annotation(  soms_AL_gene_row_annot_sum) # %>%
                            # add_col_annotation( soms_AL_gene_col_annot_sum)

cneo_soms_AL_select_annot
```

