---
title: "R Notebook"
output: html_notebook
---

## Load libraries and scripts
```{r}

if(!require(pacman)) {
  install.packages("pacman")
  library(pacman)
}

p_load(tidyverse)
p_load(GO.db)
p_load(iheatmapr)		
p_load(viridis)
p_load(readxl)

source( '../Common/global_parameters.R')
source( '../Common/helper_functions.R')

sessionInfo()

```

## Directories Management
```{r}
soms_results_dir <- file.path(results_dir, "Self_Organising_Maps/EdgeR_RUVs_QL/To_Use_k_0_Remove_5a/Condition_on_child_terms_true" ) 


soms_individual_clusters_dir <- file.path( soms_results_dir, "Individual_Clusters" )
create_dir_if_not_exists(soms_individual_clusters_dir)


```

## Read list of selected GO terms
```{r}
selected_GO_terms <- readxl::read_xlsx( file.path(data_dir, "Data_for_SOMs/Selected SOMs enriched terms.xlsx" ), sheet="Lookup_Table" ) 
  
```

## Read log fold-change of AMB and AMB-LF treatment and how they compare with each other
```{r}
compare_fold_change <- read_tsv( file = file.path( results_dir, "Compare_Fold_Change", "compare_A_AL_fold_change.tab" ) ) 

compare_fold_change_clean <- compare_fold_change %>%
                             dplyr::distinct( genes, coarse_groups) 
                             
```

## Read list of enriched GO terms related to each SOM cluster
```{r}

## Enriched GO terms per cluster
soms_A_go_results <- read_tsv( file.path(soms_results_dir, "compare_A/5x5/enriched_go_terms_SOMs.txt"))

soms_AL_go_results <- read_tsv( file.path(soms_results_dir, "compare_AL/5x5/enriched_go_terms_SOMs.txt"))

## Genes associated with enriched GO terms per cluster
soms_A_go_genes <- read_tsv( file.path(soms_results_dir, "compare_A/5x5/enriched_genes_SOMs_updated.txt")) %>%
                   mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   ) %>%
                   mutate( cluster_id = as.integer(cluster_id))

soms_AL_go_genes <- read_tsv( file.path(soms_results_dir, "compare_AL/5x5/enriched_genes_SOMs_updated.txt")) %>%
                    mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   )  %>%
                    mutate( cluster_id = as.integer(cluster_id))

## Direction of fold-change per cluster
soms_A_go_results_cleaned <- soms_A_go_results %>%
  filter( pvalue < 0.01 & ontology == "BP") %>%
  mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   ) %>%
  mutate( cluster_id = as.integer(cluster_id)) %>%
  dplyr::select( gobpid, term, cluster_id, pvalue, bh  ) %>%
  mutate( Enrichment = case_when( bh < 0.05 ~ "High",
                                  pvalue < 0.01 & bh >= 0.05 ~ "Medium",
                                  TRUE ~ "Low")) %>%
  mutate( Enrichment_Levels = case_when( bh < 0.05 ~ 2,
                                  pvalue < 0.01 & bh >= 0.05 ~ 1,
                                  TRUE ~ 0)) %>%  
  distinct()


soms_AL_go_results_cleaned <- soms_AL_go_results %>%
  filter( pvalue < 0.01 & ontology == "BP") %>%
  mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   ) %>%
  mutate( cluster_id = as.integer(cluster_id)) %>%
  dplyr::select( gobpid, term, cluster_id, pvalue, bh  ) %>%
  mutate( Enrichment = case_when( bh < 0.05 ~ "High",
                                  pvalue < 0.01 & bh >= 0.05 ~ "Medium",
                                  TRUE ~ "Low")) %>%
  mutate( Enrichment_Levels = case_when( bh < 0.05 ~ 2,
                                  pvalue < 0.01 & bh >= 0.05 ~ 1,
                                  TRUE ~ 0)) %>%  
  distinct()


```


## Read in the table showing direction of gene expression (for the genes in the cluster)
```{r}
compare_A_direction <- read_tsv ( file.path( soms_results_dir,"compare_A/5x5/soms_clusters_mapping.txt"  )) %>%
                       mutate( Direction = case_when( str_detect(Direction, "Up")  ~ "Up",
                                                      str_detect(Direction, "Down")  ~ "Down",
                                                      TRUE ~ NA_character_
                                                      ) )        
  
compare_AL_direction <- read_tsv ( file.path( soms_results_dir,"compare_AL/5x5/soms_clusters_mapping.txt"  )) %>%
                       mutate( Direction = case_when( str_detect(Direction, "Up") ~ "Up",
                                                      str_detect(Direction, "Down") ~ "Down",
                                                      TRUE ~ NA_character_
                                                      ) )        

```

## Get Codebook vector
```{r}

## Read SOMS object
compare_A_soms_obj <- readRDS ( file.path( soms_results_dir, "compare_A/5x5/soms_results.RDS" ) )
compare_AL_soms_obj <- readRDS ( file.path( soms_results_dir, "compare_AL/5x5/soms_results.RDS" ) )

## Get a list of all the clusters which actually has data in them 
compare_A_cluster_with_data <- compare_A_direction %>%
                               mutate( cluster =  purrr::map_chr( cluster, ~paste( "V", ., sep="")) ) %>%
                               pull(cluster)

compare_AL_cluster_with_data <- compare_AL_direction %>%
                               mutate( cluster =  purrr::map_chr( cluster, ~paste( "V", ., sep="")) ) %>%
                               pull(cluster)


str(compare_A_soms_obj$som$codes)

compare_A_codebook_vector_raw  <- compare_A_soms_obj$som$codes[[1]]
compare_AL_codebook_vector_raw <- compare_AL_soms_obj$som$codes[[1]]

compare_A_codebook_vector <- compare_A_codebook_vector_raw[compare_A_cluster_with_data,]
compare_AL_codebook_vector <- compare_AL_codebook_vector_raw[compare_AL_cluster_with_data,]
                 

```

## Dictionary for converting sample number in x-axis (1 to 11) to sample name (CA1, CA2, CA3, A1, A2, A3, CAL1, CAL2, CAL3, AL1, and AL3)
```{r}


if ( any( colnames( compare_A_codebook_vector )  != colnames( compare_AL_codebook_vector )  ) ) {
  
  stop ( "Code book vector sample names order for AMB treatment and AMB + LF treatment are not consistent.")
}


convert_num_to_sample_name <-      colnames( compare_A_codebook_vector ) 
names(convert_num_to_sample_name) <- seq_along(  colnames( compare_A_codebook_vector ) )


```

## Read gene per SOM cluster and the standardised gene expression value for each sample
Merge the gene per cluster information with the gene annotation information
```{r}
compare_A_soms <- read.table(file.path(soms_results_dir, "compare_A/5x5/cluster_for_each_gene.tab"), stringsAsFactors = FALSE )

compare_AL_soms <- read.table(file.path(soms_results_dir, "compare_AL/5x5/cluster_for_each_gene.tab"), stringsAsFactors = FALSE)
```


## Sample Order to use in the figure 
```{r}
types_of_treatments <- c(    "CA", "A",   "CAL", "AL")
replicate_number <- c( '1', '2', '3')

input_for_cross_df <- list (replicate_number, types_of_treatments )
names( input_for_cross_df) <- c( "replicate_number", "types_of_treatments")

all_time_points <-  purrr::cross_df( input_for_cross_df) %>%
						dplyr::mutate( sample_name = paste0( types_of_treatments, replicate_number )) %>%
						dplyr::select ( one_of ( c( "sample_name"))) %>%
						as.data.frame() %>%
						t() %>%
						as.vector()

all_time_points <- all_time_points[all_time_points != "AL2"]
```


## Need to convert SOM cluster for each gene x-axis samples number (e.g. 1, 2, 3) to sample name (CA1, CA2, CA3, A1, A2, A3, CAL1, CAL2, CAL3, AL1, and AL3)
```{r}
compare_A_soms_cleaned <- compare_A_soms %>%
                          mutate( sample_name = purrr::map_chr( x, ~convert_num_to_sample_name[.]   )  ) %>%
                          mutate( sample_name = factor( sample_name, levels = all_time_points))


compare_AL_soms_cleaned <- compare_AL_soms %>%
                           mutate( sample_name = purrr::map_chr( x, ~convert_num_to_sample_name[.]   )  ) %>%
                          mutate( sample_name = factor( sample_name, levels = all_time_points))
```

## Merge gene annotation with gene and SOM cluster information
```{r}
gene_annotation <- read_tsv( file.path(data_dir, "Gene_Annotation/cneo_gene_annotation_summary.txt" ))


compare_A_soms_annot <- compare_A_soms %>%
  dplyr::distinct( id_a, cluster) %>%
  left_join( gene_annotation, by = c( "id_a" = "orf_id")) %>%
  dplyr::distinct( id_a, cluster, is_tf, is_kinases, is_virulence_gene, order_peaktime, scer_orthologs)


compare_AL_soms_annot <- compare_AL_soms %>%
  dplyr::distinct( id_a, cluster) %>%
  left_join( gene_annotation, by = c( "id_a" = "orf_id")) %>%
  dplyr::distinct( id_a, cluster, is_tf, is_kinases, is_virulence_gene, order_peaktime, scer_orthologs)


```



## Collate data for AMB treatment's SOM cluster versus enriched GO terms heat map
```{r}

## Table of Cluster versus GO term and enrichment significance
soms_A_select_go <- soms_A_go_results_cleaned %>%
    inner_join( selected_GO_terms, by = c( "gobpid" = "GO_ID") ) %>%
    filter( Treatment == "AMB treatment" ) %>%
    arrange( Responses, GO_term  ) %>%
    dplyr::select( - Cluster) %>%
    distinct()

## Convert Cluster versus enriched GO term data into matrix format 
soms_A_select_go_matrix <- soms_A_select_go %>% 
    dplyr::select( Responses, cluster_id, term, Enrichment_Levels ) %>%
     arrange( cluster_id, Responses, term  ) %>%
    distinct() %>%
    spread( cluster_id, Enrichment_Levels, fill = NA)  %>%
    arrange( Responses, term  ) %>%
    dplyr::select( - Responses) %>%
    distinct( ) %>%
    as.data.frame %>%
    column_to_rownames("term") %>%
    as.matrix()

## Response (e.g. Stress response, ER response) versus GO terms table 
soms_A_group_term <- soms_A_select_go %>%
      distinct( Responses, GO_term  ) %>%
      arrange( Responses, GO_term  ) 

## Tidy up the direction of gene expression change for each cluster
soms_A_direction <-  soms_A_select_go %>%
                      distinct( cluster_id) %>%
                      inner_join ( compare_A_direction, by=c("cluster_id" = "cluster" ) ) %>%
                      distinct( cluster_id, Direction) %>%
                      arrange( cluster_id)
                     
if( any ( colnames( soms_A_select_go_matrix) != soms_A_direction$cluster_id )) {
  stop("Order of SOMs cluster not matched for merging with table describing the direction of gene expression change.")
}
```

## Count for each cluster, number of significantly differentially expressed genes specific to AMB treament only
```{r}

count_soms_A_only_genes <- soms_A_go_genes %>%
  left_join( compare_fold_change_clean, by = c( "gene_id" = "genes" )) %>%
  distinct( gene_id, cluster_id, coarse_groups) %>%
  group_by(cluster_id, coarse_groups) %>%
  summarise( counts = n()) %>%
  ungroup %>%
  spread( coarse_groups, counts, fill = 0) %>%
  dplyr::select( -Same) %>%
  ## Add the clusters that did not have any significant
  ## differentially expressed genes back into the table
  right_join ( compare_A_direction, 
               by = c( "cluster_id" = "cluster" )  ) %>%
  mutate( `AMB only` = ifelse(is.na(`AMB only`), 0, `AMB only`)) %>%
  ## Filter the table so only the clusters associated with selected GO terms are in the table
  filter( cluster_id %in% colnames( soms_A_select_go_matrix)) %>%
  dplyr::select( cluster_id, Direction, `AMB only`) %>%
  dplyr::rename( `AMB specific` = "AMB only") %>%
  arrange( cluster_id ) %>%
  column_to_rownames("cluster_id")


```

## Plot AMB only treatment SOM cluster vs. selected enriched GO terms using iheatmapr
```{r}
grid_params <- setup_colorbar_grid(x_spacing = 0.3)


cneo_soms_A_select <- main_heatmap(  soms_A_select_go_matrix, colors= c("lightblue", "darkblue"), 
                                     layout = list(margin = list(l = 350)),
                                     name = "Significance",
                                      colorbar_grid = grid_params )  %>%
  add_row_labels() %>%
  add_col_labels() %>%
  add_row_annotation( data.frame( Responses =  soms_A_group_term$Responses)) %>%
  add_col_annotation( count_soms_A_only_genes, colors = list(Direction=c("green", "red")))

cneo_soms_A_select

dim(soms_A_select_go_matrix)

save_iheatmap(p=cneo_soms_A_select, filename=file.path( soms_results_dir, "compare_A/5x5/compare_A_selected_go_terms_iheatmapr.png"))

```

## AMB + LF treatment
```{r}

## Table of Cluster versus GO term and enrichment significance
soms_AL_select_go <- soms_AL_go_results_cleaned %>%
    inner_join( selected_GO_terms, by = c( "gobpid" = "GO_ID") ) %>%
    filter( Treatment == "AMB-LF treatment" ) %>%
    arrange( Responses, GO_term  ) %>%
    dplyr::select( - Cluster) %>%
    distinct()

## Convert Cluster versus enriched GO term data into matrix format 
soms_AL_select_go_matrix <- soms_AL_select_go %>% 
    dplyr::select( Responses, cluster_id, term, Enrichment_Levels ) %>%
     arrange( cluster_id, Responses, term  ) %>%
    distinct() %>%
    spread( cluster_id, Enrichment_Levels, fill = NA)  %>%
    arrange( Responses, term  ) %>%
    dplyr::select( - Responses) %>%
    distinct( ) %>%
    as.data.frame %>%
    column_to_rownames("term") %>%
    as.matrix()

## Response (e.g. Stress response, ER response) versus GO terms table 
soms_AL_group_term <- soms_AL_select_go %>%
      distinct( Responses, GO_term  ) %>%
      arrange( Responses, GO_term  ) 

## Tidy up the direction of gene expression change for each cluster
soms_AL_direction <-  soms_AL_select_go %>%
                      distinct( cluster_id) %>%
                      inner_join ( compare_AL_direction, by=c("cluster_id" = "cluster" ) ) %>%
                      distinct( cluster_id, Direction) %>%
                      arrange( cluster_id)
                     
if( any ( colnames( soms_AL_select_go_matrix) != soms_AL_direction$cluster_id )) {
  stop("Order of SOMs cluster not matched for merging with table describing the direction of gene expression change.")
}
```

## Count for each cluster, number of significantly differentially expressed genes specific to AMB-LF treatment only
```{r}
count_soms_AL_only_genes <- soms_AL_go_genes %>%
  left_join( compare_fold_change_clean, by = c( "gene_id" = "genes" )) %>%
  distinct( gene_id, cluster_id, coarse_groups) %>%
  group_by(cluster_id, coarse_groups) %>%
  summarise( counts = n()) %>%
  ungroup %>%
  spread( coarse_groups, counts, fill = 0) %>%
  dplyr::select( -Same) %>%
  ## Add the clusters that did not have any significant
  ## differentially expressed genes back into the table  
  right_join ( compare_AL_direction, 
               by = c( "cluster_id" = "cluster" )  ) %>%
  mutate( `AMB-LF only` = ifelse(is.na(`AMB-LF only`), 0, `AMB-LF only`)) %>%
  ## Filter the table so only the clusters associated with selected GO terms are in the table
  filter( cluster_id %in% colnames( soms_AL_select_go_matrix)) %>%
  dplyr::select( cluster_id, Direction, `AMB-LF only`) %>%
  dplyr::rename( `AMB-LF specific` = "AMB-LF only") %>%
  arrange( cluster_id ) %>%
  column_to_rownames("cluster_id")

# soms_AL_go_genes
# compare_fold_change_clean

```

## Plot AMB-LF treatment SOM cluster vs. selected enriched GO terms using iheatmapr
```{r}


grid_params <- setup_colorbar_grid(x_spacing = 0.3)

cneo_soms_AL_select <- main_heatmap(  soms_AL_select_go_matrix, colors= c("lightblue", "darkblue"), 
                                      layout = list(margin = list(l = 350)),
                                      name = "Significance",
                                      colorbar_grid = grid_params )  %>%
  add_row_labels() %>%
  add_col_labels() %>%
  add_row_annotation( data.frame( Responses =  soms_AL_group_term$Responses)) %>%
  add_col_annotation( count_soms_AL_only_genes, colors = list(Direction=c("green", "red")))

cneo_soms_AL_select

save_iheatmap(p=cneo_soms_AL_select, filename=file.path( soms_results_dir, "compare_AL/5x5/compare_AL_selected_go_terms_iheatmapr.png"))

dim(soms_AL_select_go_matrix)
```

## Try to add more annotation, AMB only
```{r}

soms_A_gene_annotation <- soms_A_select_go %>%
  inner_join  ( soms_A_go_genes, by =c( "cluster_id" = "cluster_id",
                                       "gobpid" = "go_id") ) %>%
  dplyr::distinct( gobpid, term, cluster_id, Treatment, gene_id, is_tf, is_kinases, is_virulence_gene, order_peaktime)

soms_A_gene_row_annot_sum <- soms_A_gene_annotation %>%
  dplyr::select( -cluster_id, -Treatment) %>%
  dplyr::distinct( ) %>%
  group_by( term ) %>%
  summarise( # count_tfs =   sum(ifelse(is.na(is_tf), 0, 1)),
             count_kinases =   sum(ifelse(is.na(is_kinases), 0, 1)),
             count_virulence_genes = sum(ifelse(is.na(is_virulence_gene), 0, 1)),
             count_cell_cycle_genes = sum(ifelse(is.na(order_peaktime), 0, 1)))  %>%
 arrange(term) %>%
 column_to_rownames( "term")

dim(soms_A_select_go_matrix)
dim(soms_A_gene_row_annot_sum )


soms_A_gene_row_annot_sum <- soms_A_gene_row_annot_sum[rownames(soms_A_select_go_matrix), ]

rownames( soms_A_gene_row_annot_sum) == rownames(soms_A_select_go_matrix)

cneo_soms_A_select_annot <- cneo_soms_A_select %>%
                            add_row_annotation(  soms_A_gene_row_annot_sum) 


cneo_soms_A_select_annot


save_iheatmap(p=cneo_soms_A_select_annot, 
              filename=file.path( soms_results_dir, 
                                  "compare_A/5x5/compare_A_selected_go_terms_iheatmapr_annot.png"))

```

## Try to add more annotation, AMB-LF

iron ion transmembrane transport, GO:0034755, AMB-LF treatment Cluster 10, have 1 virulence gene (CNAG_03464 /  Cu-oxidase / Laccase-2 / EC 1.10.3.2 / Diphenol oxidase 2). 
```{r}



soms_AL_gene_annotation <- soms_AL_select_go %>%
  inner_join  ( soms_AL_go_genes, by =c( "cluster_id" = "cluster_id",
                                       "gobpid" = "go_id") ) %>%
  dplyr::distinct( gobpid, term, cluster_id, Treatment, gene_id, is_tf, is_kinases, is_virulence_gene, order_peaktime)

soms_AL_gene_row_annot_sum <- soms_AL_gene_annotation %>%
  dplyr::select( -cluster_id, -Treatment ) %>%
  dplyr::distinct( ) %>%
  group_by( term ) %>%
  summarise(  #  count_tfs =   sum(ifelse(is.na(is_tf), 0, 1)),
             count_kinases =   sum(ifelse(is.na(is_kinases), 0, 1)),
             count_virulence_genes = sum(ifelse(is.na(is_virulence_gene), 0, 1)),
             count_cell_cycle_genes = sum(ifelse(is.na(order_peaktime), 0, 1))) %>%
 arrange(term) %>%
 column_to_rownames( "term")


soms_AL_gene_row_annot_sum <- soms_AL_gene_row_annot_sum[rownames( soms_AL_select_go_matrix), ]

rownames( soms_AL_gene_row_annot_sum) == rownames( soms_AL_select_go_matrix)


cneo_soms_AL_select_annot <- cneo_soms_AL_select %>%
                            add_row_annotation(  soms_AL_gene_row_annot_sum) 

cneo_soms_AL_select_annot


save_iheatmap(p=cneo_soms_AL_select_annot, 
              filename=file.path( soms_results_dir,
                                  "compare_AL/5x5/compare_AL_selected_go_terms_iheatmapr_annot.png"))

```


```{r}
colnames( soms_AL_select_go)
colnames( soms_AL_go_genes)
colnames( compare_fold_change)
colnames( compare_AL_soms_cleaned)

```


## Plotting Meta Heat Map
```{r}

soms_AL_select_go_and_gene <- soms_AL_select_go %>% 
  inner_join(  soms_AL_go_genes, by = c( "gobpid" = "go_id",
                                        "cluster_id" = "cluster_id") ) %>%
  inner_join( compare_fold_change, by=c("gene_id" = "genes") )  %>%
  inner_join( compare_AL_soms_cleaned, by =c( "gene_id" = "id_a",
                                              "cluster_id" = "cluster")) %>%
  dplyr::select(one_of( c( ## The IDs
                           "cluster_id",  "gene_id", "gobpid", "term", "ontology",
                           ## Gene Ontology enrichment Information 
                           "pvalue", "bh", "Enrichment", "Enrichment_Levels",  "Responses", 
                           ## Cluster level data 
                           "Treatment", "Regulation/p-value",  
                           ## Differential Gene Expression level data
                           "gene_names", "fungidb_gene_name", "logFC.AL", "FDR.AL", "groups", "coarse_groups",
                           ## Standardized Gene Expression Level data
                           "sample_name", "y")))  

## Do one plot for each cluster
my_input_go_and_gene <- soms_AL_select_go_and_gene
my_cluster <- 1
my_log_fc_name <- quo( logFC.AL ) 


## Plot Standardized Gene Expression Levels
 soms_AL_std_gene_exp <- my_input_go_and_gene %>%
        ## do this for cluster 1 first 
        filter ( cluster_id == my_cluster) %>%
        dplyr::distinct( cluster_id, gene_id, sample_name, y ) %>%
        tidyr::spread( sample_name, y) %>%
        arrange(cluster_id, gene_id) %>%
        dplyr::select(-cluster_id ) %>%
        column_to_rownames("gene_id") %>%
        as.matrix
  
dim(soms_AL_std_gene_exp)

## Plot log fold-change per gene 
soms_AL_log_fc <- my_input_go_and_gene %>%
        ## do this for cluster 1 first 
        filter ( cluster_id == my_cluster) %>%
        dplyr::distinct( cluster_id, gene_id, !!my_log_fc_name)  %>%
        arrange(cluster_id, gene_id) %>%
        dplyr::select( -cluster_id, -gene_id) %>%
        dplyr::rename( !! sym("AMB-LF log FC") := quo_name( my_log_fc_name ))


## Plot gene ontology terms 
soms_AL_go_terms <- my_input_go_and_gene %>%
        ## do this for cluster 1 first 
        filter ( cluster_id == my_cluster) %>%
        dplyr::distinct( cluster_id, gene_id, term, Enrichment_Levels) %>%
        tidyr::spread( term, Enrichment_Levels, fill = 0) %>%
        arrange(cluster_id, gene_id) %>%
        dplyr::select( -cluster_id) %>%
        column_to_rownames("gene_id") %>%
        as.matrix()

## Get_gene_name 
soms_AL_gene_names <- my_input_go_and_gene %>%
        ## do this for cluster 1 first 
        filter ( cluster_id == my_cluster) %>%
        dplyr::distinct( cluster_id, gene_id, gene_names, fungidb_gene_name)   %>%
        dplyr::mutate( tick_text = paste( ifelse( is.na(gene_names), "", gene_names), gene_id, fungidb_gene_name)) %>%
        arrange(cluster_id, gene_id)


# soms_AL_gene_description <- my_input_go_and_gene %>%
#         ## do this for cluster 1 first 
#         filter ( cluster_id == my_cluster) %>%
#         dplyr::distinct( cluster_id, gene_id, fungidb_gene_name)   %>%
#         arrange(cluster_id, gene_id)


my_margins <- list(
  l = 15,
  r = 5,
  b = 10, # 10
  t = 5,
  pad = 2
)

plot_iheatmapr <- main_heatmap(soms_AL_std_gene_exp, layout = list(margin = my_margins)) %>%
  add_col_labels(size = max(sapply( colnames(soms_AL_go_terms), str_length ))/65) %>%
  add_row_annotation(soms_AL_log_fc) %>%
  add_row_labels( ticktext= soms_AL_gene_names$tick_text, size = max(sapply(soms_AL_gene_names$tick_text, str_length ))/40 ) %>%
  add_main_heatmap(soms_AL_go_terms, 
                   name= "Enrichemtn Levels", 
                   colors =c("white", "lightblue", "darkblue"),
                   size = ncol( soms_AL_go_terms)/((ncol(soms_AL_std_gene_exp) + 1.5)*1.5) ) %>%
  add_col_labels(size = max(sapply( colnames(soms_AL_go_terms), str_length ))/65) # %>%
  # add_row_labels( ticktext = soms_AL_gene_description$fungidb_gene_name, side = "right")


save_iheatmap(p=plot_iheatmapr, filename=file.path( soms_individual_clusters_dir, "som_AL_cluster_1_go_terms.pdf" ))
save_iheatmap(p=plot_iheatmapr, filename=file.path( soms_individual_clusters_dir, "som_AL_cluster_1_go_terms.png" ))

```


## 
```{r}







```















