---
title: "R Notebook"
output: html_notebook
---


```{r}

if(!require(pacman)) {
  install.packages("pacman")
  library(pacman)
}

p_load(tidyverse)
p_load(gplots)
p_load(iheatmapr)


source( '../../Common/global_parameters.R')

```


## Global parameters
```{r}
heatmap_column_labels <-  c( "CA1", "CA2", "CA3", "A1", "A2", "A3", "CAL1", "CAL2", "CAL3", "AL1",  "AL3" ) 

```

## Read in input files
```{r}

soms_results_dir <- file.path( results_dir, "Self_Organising_Maps/EdgeR_RUVs_QL/To_Use_k_0_Remove_5a/Condition_on_child_terms_true" ) 

compare_A_soms <- readRDS ( file.path( soms_results_dir, "compare_A/5x5/soms_results.RDS" ) )
compare_AL_soms <- readRDS ( file.path( soms_results_dir, "compare_AL/5x5/soms_results.RDS" ) )

## Read in the table showing direction of gene expression (for the genes in the cluster)
compare_A_direction <- read_tsv ( file.path( soms_results_dir,"compare_A/5x5/soms_clusters_mapping.txt"  )) %>%
                       mutate( Direction = case_when( str_detect(Direction, "Up")  ~ "Up",
                                                      str_detect(Direction, "Down")  ~ "Down",
                                                      TRUE ~ NA_character_
                                                      ) )        
  
compare_AL_direction <- read_tsv ( file.path( soms_results_dir,"compare_AL/5x5/soms_clusters_mapping.txt"  )) %>%
                       mutate( Direction = case_when( str_detect(Direction, "Up") ~ "Up",
                                                      str_detect(Direction, "Down") ~ "Down",
                                                      TRUE ~ NA_character_
                                                      ) )        

## Get a list of all the clusters which actually has data in them 
compare_A_cluster_with_data <- compare_A_direction %>%
                               mutate( cluster =  purrr::map_chr( cluster, ~paste( "V", ., sep="")) ) %>%
                               pull(cluster)

compare_AL_cluster_with_data <- compare_AL_direction %>%
                               mutate( cluster =  purrr::map_chr( cluster, ~paste( "V", ., sep="")) ) %>%
                               pull(cluster)


str(compare_A_soms$som$codes)

compare_A_codebook_vector_raw  <- compare_A_soms$som$codes[[1]]
compare_AL_codebook_vector_raw <- compare_AL_soms$som$codes[[1]]

compare_A_codebook_vector <- compare_A_codebook_vector_raw[compare_A_cluster_with_data,]
compare_AL_codebook_vector <- compare_AL_codebook_vector_raw[compare_AL_cluster_with_data,]
                 

## Remove V's from Cluster ID
rownames(compare_A_codebook_vector)  <- str_replace( rownames(compare_A_codebook_vector), "V", "")
rownames(compare_AL_codebook_vector)  <- str_replace( rownames(compare_AL_codebook_vector), "V", "")


## Genes associated with enriched GO terms per cluster
soms_A_go_genes <- read_tsv( file.path(soms_results_dir, "compare_A/5x5/enriched_genes_SOMs_updated.txt")) %>%
                   mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   ) %>%
                   mutate( cluster_id = as.integer(cluster_id))

soms_AL_go_genes <- read_tsv( file.path(soms_results_dir, "compare_AL/5x5/enriched_genes_SOMs_updated.txt")) %>%
                    mutate( cluster_id = str_replace( query_list_id, "SOMs_cluster_C", "")   )  %>%
                    mutate( cluster_id = as.integer(cluster_id))

## Read log fold-change of AMB and AMB+LF treatment and how they compare with each other
compare_fold_change <- read_tsv( file = file.path( results_dir, "Compare_Fold_Change", "compare_A_AL_fold_change.tab" ) ) 

compare_fold_change_clean <- compare_fold_change %>%
                             dplyr::distinct( genes, coarse_groups) 
                             

```




## Check consistency of code book vector sample names order
```{r}
if ( any( colnames( compare_A_codebook_vector )  != colnames( compare_AL_codebook_vector )  ) ) {
  
  stop ( "Code book vector sample names order for AMB treatment and AMB + LF treatment are not consistent.")
}


```


```{r}
##################################################################################################################

# Heatmap of the codebook vector

my_colour_palette <- colorRampPalette(c("#2166ac","#f7f7f7","#b2182b"))(75) 
margins_size_factor <- 5
texts_size_factor_col <- 10
texts_size_factor_row <- 30

### AMB only model, A - CA
pdf( file.path( soms_results_dir, "compare_A/5x5/edgeR_compare_A_som_codebook_vector.pdf") ) 

heatmap.2(compare_A_codebook_vector[,heatmap_column_labels],  symm = FALSE, col = my_colour_palette, Colv=NA,  #distfun = function(c) as.dist(1 - c),
          dendrogram = "row", hclustfun=function(h)hclust(h,method="average"),  density.info="none", trace="none", # main = "Clustering miRNA with low to medium CV",
          #labRow=,
          xlab="Samples",
          ylab="Clusters",
          labCol= heatmap_column_labels, 
          cexCol= 0.0 + 1/log(length(compare_A_codebook_vector[1,]), texts_size_factor_col), cexRow= 0.0 + 1/log(length(compare_A_codebook_vector[,1]), texts_size_factor_row),
          margins = c(margins_size_factor, margins_size_factor),
          offsetRow = 0.02,
          offsetCol = 0.02 )
dev.off()


### AMB + LF model, AL - CAL
pdf( file.path( soms_results_dir, "compare_AL/5x5/edgeR_compare_AL_som_codebook_vector.pdf") ) 
          
heatmap.2(compare_AL_codebook_vector[,heatmap_column_labels],  symm = FALSE, col = my_colour_palette, Colv=NA,  #distfun = function(c) as.dist(1 - c),
          dendrogram = "row", hclustfun=function(h)hclust(h,method="average"),  density.info="none", trace="none", # main = "Clustering miRNA with low to medium CV",
          #labRow=, 
          xlab="Samples",
          ylab="Clusters",
          labCol= heatmap_column_labels, 
          cexCol= 0.0 + 1/log(length(compare_AL_codebook_vector[1,]), texts_size_factor_col), cexRow= 0.0 + 1/log(length(compare_AL_codebook_vector[,1]), texts_size_factor_row),
          margins = c(margins_size_factor, margins_size_factor),
          offsetRow = 0.02,
          offsetCol = 0.02 )
dev.off()
          


##################################################################################################################


```

## Count for each cluster, number of significantly differentially expressed genes specific to AMB treament or AMB+LF treatment only
```{r}

count_soms_A_only_genes <- soms_A_go_genes %>%
 left_join( compare_fold_change_clean, by = c( "gene_id" = "genes" )) %>%
  distinct( gene_id, cluster_id, coarse_groups) %>%
  group_by(cluster_id, coarse_groups) %>%
  summarise( counts = n()) %>%
  ungroup %>%
  spread( coarse_groups, counts, fill = 0) %>%
  dplyr::select( -Same) %>%
  right_join ( data.frame( cluster_id = as.integer(rownames( compare_A_codebook_vector[,heatmap_column_labels]))), 
               by = "cluster_id") %>%
  mutate( `AMB only` = ifelse(is.na(`AMB only`), 0, `AMB only`))


count_soms_AL_only_genes <- soms_AL_go_genes %>%
 left_join( compare_fold_change_clean, by = c( "gene_id" = "genes" )) %>%
  distinct( gene_id, cluster_id, coarse_groups) %>%
  group_by(cluster_id, coarse_groups) %>%
  summarise( counts = n()) %>%
  ungroup %>%
  spread( coarse_groups, counts, fill = 0) %>%
  dplyr::select( -Same, -Opposite) %>%
  right_join ( data.frame( cluster_id = as.integer(rownames( compare_AL_codebook_vector[,heatmap_column_labels]))), 
               by = "cluster_id") %>%
  mutate( `AMB+LF only` = ifelse(is.na(`AMB+LF only`), 0, `AMB+LF only`))

# soms_AL_go_genes
# compare_fold_change_clean

```

## Draw the heat map using iheatmapr
```{r}

my_margins <- list(
  l = 5,
  r = 5,
  b = 10, # 10
  t = 5,
  pad = 2
)

my_font <- list(
  family = "sans serif",
  size = 16,
  color = 'black')

my_vwidth <- 6
my_vheight <- 7 # 7

dim(compare_A_codebook_vector[,heatmap_column_labels])

### AMB only model, A - CA

compare_A_row_annot <- data.frame( Direction =  compare_A_direction$Direction,
                                   AMB_only = count_soms_A_only_genes$`AMB only`)

colnames( compare_A_row_annot) <- c("Direction", "AMB<br>specific")


compare_A_iheatmapr <- main_heatmap( compare_A_codebook_vector[,heatmap_column_labels], 
                                     layout = list(margin = my_margins, font=my_font),
                                     name = "Standardized<br>Gene<br>Expression") %>%
  add_row_labels(size = 0.1) %>%
  add_col_labels() %>%
  add_row_clustering() %>%
  add_col_title("Clusters", buffer = 0.03) %>%
  add_row_title ("Samples",  buffer = 0.01) %>%
  add_row_annotation(  compare_A_row_annot ) 

compare_A_iheatmapr

save_iheatmap( p=compare_A_iheatmapr, 
               filename = file.path( soms_results_dir, "compare_A/5x5/edgeR_compare_A_som_codebook_vector_iheatmapr.pdf"),
               vwidth = my_vwidth, vheight = my_vheight  )
save_iheatmap( p=compare_A_iheatmapr, 
               filename = file.path( soms_results_dir, "compare_A/5x5/edgeR_compare_A_som_codebook_vector_iheatmapr.png"),
               vwidth = my_vwidth, vheight = my_vheight  )


### AMB + LF model, AL - CAL

compare_AL_row_annot <- data.frame( Direction =  compare_AL_direction$Direction,
                                   AMB_LF_only = count_soms_AL_only_genes$`AMB+LF only`)

colnames( compare_AL_row_annot) <- c("Direction", "AMB+LF<br>specific")

compare_AL_iheatmapr <- main_heatmap( compare_AL_codebook_vector[,heatmap_column_labels], 
                                     layout = list(margin = my_margins, font=my_font),
                                     name = "Standardized<br>Gene<br>Expression" )  %>%
  add_row_labels(size = 0.1) %>%
  add_col_labels() %>%
  add_row_clustering() %>%
  add_col_title("Clusters", buffer = 0.03) %>%
  add_row_title ("Samples",  buffer = 0.01) %>%
  add_row_annotation(  compare_AL_row_annot)


compare_AL_iheatmapr

save_iheatmap( p=compare_AL_iheatmapr, 
               filename = file.path( soms_results_dir, "compare_AL/5x5/edgeR_compare_AL_som_codebook_vector_iheatmapr.pdf"),
               vwidth = my_vwidth, vheight = my_vheight     )
save_iheatmap( p=compare_AL_iheatmapr, 
               filename = file.path( soms_results_dir, "compare_AL/5x5/edgeR_compare_AL_som_codebook_vector_iheatmapr.png"),
               vwidth = my_vwidth, vheight = my_vheight)


```




