---
title: "R Notebook"
output: html_notebook
---



```{r}

if(!require(pacman)) {
  install.packages("pacman")
  library(pacman)
}

p_load(tidyverse)
p_load(plotly)
p_load(UpSetR)
p_load(RColorBrewer)

source( '../Common/global_parameters.R')

```

## Read in input files
```{r}

de_genes_dir <- file.path( results_dir, "Differential_Expression/EdgeR_RUVs_QL/cneo/To_Use_k_0_Remove_5a/Gene_Lists")

compare_A_de_genes <- read_tsv( file.path(de_genes_dir, "compare_A.tab" )) 
compare_AL_de_genes <- read_tsv( file.path(de_genes_dir, "compare_AL.tab" ))

compare_A_de_genes_clean <- compare_A_de_genes %>%
                            dplyr::select( genes, logFC, FDR)

compare_AL_de_genes_clean <- compare_AL_de_genes %>%
                            dplyr::select( genes, logFC, FDR)

```


## Get gene annotation
```{r}
gene_annotation   <- read_tsv( file.path( data_dir, "Gene_Annotation/cneo_gene_annotation_summary.txt"  )  )


```


## Merge the AMB and AMB + LF treatments gene expressions fold-change table. 
```{r}
# dim( compare_A_de_genes)
# dim( compare_AL_de_genes)

compare_fold_change <- compare_A_de_genes_clean %>%
  inner_join( compare_AL_de_genes_clean, by = "genes", suffix = c( ".A", ".AL") ) %>%
  mutate( groups = case_when( FDR.A < 0.05 & FDR.AL < 0.05 & sign(logFC.A) != sign(logFC.AL) & logFC.A > 0  ~ "AMB up, AMB+LF down",
                              FDR.A < 0.05 & FDR.AL < 0.05 & sign(logFC.A) != sign(logFC.AL) & logFC.AL > 0  ~ "AMB down, AMB+LF up",
                              FDR.A < 0.05 & FDR.AL < 0.05 & sign(logFC.A) == sign(logFC.AL) & logFC.A > 0 ~ "Both up",
                              FDR.A < 0.05 & FDR.AL < 0.05 & sign(logFC.A) == sign(logFC.AL) & logFC.A < 0 ~ "Both down",
                              FDR.A < 0.05 & FDR.AL >= 0.05 & logFC.A > 0 ~ "AMB up only",
                              FDR.A < 0.05 & FDR.AL >= 0.05 & logFC.A < 0 ~ "AMB down only",
                              FDR.A >= 0.05 & FDR.AL < 0.05 & logFC.AL > 0 ~ "AMB+LF up only",
                              FDR.A >= 0.05 & FDR.AL < 0.05 & logFC.AL < 0 ~ "AMB+LF down only",
                              TRUE ~ "Not significant" )    ) %>%
  mutate( groups = factor (groups, levels=c( "AMB up, AMB+LF down", 
                                             "AMB down, AMB+LF up",
                                             "Both up",
                                             "Both down",
                                             "AMB up only",
                                             "AMB down only",
                                             "AMB+LF up only",
                                             "AMB+LF down only",
                                             "Not significant"
                                            ))) %>%
  mutate( coarse_groups = case_when( FDR.A < 0.05 & FDR.AL < 0.05 & sign(logFC.A) != sign(logFC.AL)  ~ "Opposite",
                              FDR.A < 0.05 & FDR.AL < 0.05 & sign(logFC.A) == sign(logFC.AL)  ~ "Same",
                              FDR.A < 0.05 & FDR.AL >= 0.05 ~ "AMB only",
                              FDR.A >= 0.05 & FDR.AL < 0.05 ~ "AMB+LF only",
                              TRUE ~ "Not significant" )    ) %>%
  mutate( coarse_groups = factor( coarse_groups, levels = c( "Same", "AMB only", "AMB+LF only", "Opposite", "Not significant")))

write.table( compare_fold_change, file = file.path( results_dir, "Compare_Fold_Change", "compare_A_AL_fold_change.tab" ), 
             sep="\t", row.names = FALSE, quote = FALSE ) 

```

```{r}
scatter_plot <-  compare_fold_change  %>%
                 ggplot(aes(logFC.A, logFC.AL, col=groups)) +
                 geom_point() +
                 xlab ( "AMB treatment log fold-change") +
                 ylab ( "AMB+LF treatment log fold-change") +
                 labs( col = "Is Statistically Significant")

scatter_plot

```

There were 3 genes that had opposite sig. log fold-change in AMB and AMB+LF treatment
CNAG_02691
CNAG_05638
CNAG_04363 up in AMB only, down in AMB+LF. 
All three genes were hypothetical only. Therefore we focused on genes that are significantly differentially expressed in AMB treatment or AMB+LF treatment only. 


```{r}
coarse_scatter_plot <-  compare_fold_change  %>%
                 ggplot(aes(logFC.A, logFC.AL, col=coarse_groups, text=genes)) +
                 geom_point() +
                 facet_grid ( . ~ coarse_groups) +
                 xlab ( "AMB treatment log fold-change") +
                 ylab ( "AMB+LF treatment log fold-change") +
                 labs( col = "Is Statistically Significant")

ggplotly( coarse_scatter_plot )

ggsave( plot=coarse_scatter_plot, filename= file.path( results_dir, "Compare_Fold_Change", "compare_A_AL_fold_change.pdf"))
ggsave( plot=coarse_scatter_plot, filename= file.path( results_dir, "Compare_Fold_Change", "compare_A_AL_fold_change.png"))
ggsave( plot=coarse_scatter_plot, filename= file.path( results_dir, "Compare_Fold_Change", "compare_A_AL_fold_change.svg"))
ggsave( plot=coarse_scatter_plot, filename= file.path( results_dir, "Compare_Fold_Change", "compare_A_AL_fold_change.tiff"))

```

## Draw venn diagram and upset plot
```{r}

venn_diagram_data <- compare_fold_change %>%
  distinct() %>%
  mutate( AMB_treatment = case_when( 
                                     FDR.A < 0.05 & logFC.A > 0 ~ 'AMB treatment, up',
                                     FDR.A < 0.05 & logFC.A < 0 ~ 'AMB treatment, down',
                                     TRUE ~ 'AMB treatment, n.s.'
                                     ),
          AMB_LF_treatment = case_when( 
                                     FDR.AL < 0.05 & logFC.AL > 0 ~ 'AMB+LF treatment, up',
                                     FDR.AL < 0.05 & logFC.AL < 0 ~ 'AMB+LF treatment, down',
                                     TRUE ~ 'AMB+LF treatment, n.s.'
                                     ) )  %>%
  mutate( AMB_treatment = factor(AMB_treatment, levels=c('AMB treatment, up',  
                                                         'AMB treatment, n.s.', 
                                                         'AMB treatment, down')),
          AMB_LF_treatment = factor(AMB_LF_treatment, levels=c('AMB+LF treatment, up',  
                                                               'AMB+LF treatment, n.s.', 
                                                               'AMB+LF treatment, down') ) )  %>%
  filter( !( AMB_treatment == 'AMB treatment, n.s.' &  AMB_LF_treatment == 'AMB+LF treatment, n.s.') ) %>%
  dplyr::mutate( id = row_number() ) %>%
  dplyr::select( id, genes, AMB_treatment, AMB_LF_treatment) 
```


```{r}
venn_diagram_data_pivot <- venn_diagram_data %>%
  mutate ( value = 1) %>%
  spread( AMB_treatment, value, fill =  0) %>%
  mutate ( value = 1) %>%
  spread( AMB_LF_treatment, value, fill =  0) %>%
  dplyr::select( -one_of(c( "AMB treatment, n.s.", "AMB+LF treatment, n.s."))) %>% 
  dplyr::select( - id) %>%
  dplyr::group_by(genes) %>%
  dplyr::summarise_at( vars(contains("treatment")), sum ) %>%
  dplyr::mutate_at( vars(contains("treatment")), as.integer ) %>%
  ungroup %>%
  mutate( genes = factor(genes)) %>%
  as.data.frame()

upset_plot_queries_statement <-list(
        list(query = intersects, params = list('AMB treatment, up'), color = "#377EB8", active = T),
        list(query = intersects, params = list('AMB treatment, down'), color = "#984EA3", active = T),
        list(query = intersects, params = list('AMB+LF treatment, up'), color = "#FF7F00", active = T),
        list(query = intersects, params = list('AMB+LF treatment, down'), color = "#A65628", active = T)
        )

upset(venn_diagram_data_pivot, 
      order.by = "freq", 
      sets=rev(names(venn_diagram_data_pivot)[2:ncol(venn_diagram_data_pivot)]), 
      nsets = length( names(venn_diagram_data_pivot)), 
      keep.order=FALSE )

upset(venn_diagram_data_pivot, 
      order.by = "freq", 
      sets=rev(names(venn_diagram_data_pivot)[2:ncol(venn_diagram_data_pivot)]), 
      nsets = length( upset_input), 
      keep.order=FALSE,
      mainbar.y.label = "No. of Genes", 
      sets.x.label = "No. of Genes",
    sets.bar.color = c(   "springgreen4", "springgreen4", "red", "red" )  ,
    queries = upset_plot_queries_statement
    )
  
```



```{r}
pdf( file.path(  results_dir, "Compare_Fold_Change", "upset_plot.pdf"), width = 6.7, height = 3)
upset(venn_diagram_data_pivot, 
      order.by = "freq", 
      sets=rev(names(venn_diagram_data_pivot)[2:ncol(venn_diagram_data_pivot)]), 
      nsets = length( upset_input), 
      keep.order=FALSE,
      mainbar.y.label = "No. of Genes", 
      sets.x.label = "No. of Genes",
    sets.bar.color = c(   "springgreen4", "springgreen4", "red", "red" )  ,
    queries = upset_plot_queries_statement
    )
dev.off()
```


## Venn Diagram represented as a table 
```{r}



venn_diagram_table <- venn_diagram_data %>%
  group_by ( AMB_treatment, AMB_LF_treatment ) %>%
  count() %>%
  spread( AMB_LF_treatment, n)

venn_diagram_table

```



## Top 10 bottom 10 genes 
```{r}

all_genes_specific_to_AL <- compare_fold_change %>% 
  filter ( coarse_groups == "AMB+LF only") %>%
  arrange( desc(logFC.AL )) %>%
  mutate( rank_from_biggest_increase = row_number()) %>%
  arrange( logFC.AL ) %>%
  mutate( rank_from_biggest_decrease = row_number()) %>%
  left_join( gene_annotation, by=c("genes" = "orf_id")) %>%
  dplyr::select( -logFC.A, -FDR.A  )

write.table( all_genes_specific_to_AL, file.path( results_dir, "Compare_Fold_Change", "all_genes_specific_to_AL.tab"  ),
             row.names = FALSE, quote = FALSE, sep="\t") 


bottom_ten_log_fc_AL_specific <- compare_fold_change %>% 
  filter ( coarse_groups == "AMB+LF only") %>%
  arrange( logFC.AL ) %>%
  mutate( rank_from_biggest_decrease = row_number()) %>%
  left_join( gene_annotation, by=c("genes" = "orf_id")) %>%
  filter( rank_from_biggest_decrease <= 10 )


top_ten_log_fc_AL_specific <-  compare_fold_change %>% 
  filter ( coarse_groups == "AMB+LF only") %>%
  arrange( desc(logFC.AL )) %>%
  mutate( rank_from_biggest_increase = row_number()) %>%
  left_join( gene_annotation, by=c("genes" = "orf_id")) %>%
  filter( rank_from_biggest_increase <= 10 )


```


