---
title: "edgeR_differential_gene_expression_ruv.Rmd"
output:
  html_document: default
  html_notebook: default
---

## Load libraries and scripts
```{r}

if(!require(pacman)) {
  install.packages("pacman")
  library(pacman)
}

p_load(edgeR)
p_load(lazyeval)
p_load(tidyverse)

source( '../Common/edgeR_common_functions.R')
source( '../../../Common/global_parameters.R')
source( '../../../Common/helper_functions.R')

sessionInfo()

```

## Command-line parameters
```{r}

command_line_args <-  commandArgs(trailingOnly=T)

print( command_line_args)


if ( length(command_line_args) > 0   ) {
##### This is a parameter to change #### *******************
num_unwanted_variables <- as.integer( command_line_args[2]    ) 

##### This is a parameter to change #### *******************
is_quasi_likelihood_test <- as.logical( command_line_args[3]  )

##### This is a parameter to change #### *******************
type_of_ruv <-  as.character( command_line_args[4]  ) # RUVg, RUVs, RUVr

##### This is a parameter to change #### *******************
organism <-  as.character( command_line_args[5]  ) # RUVg, RUVs, RUVr
		
} else {
	
	num_unwanted_variables <- 1
	
	is_quasi_likelihood_test <- TRUE
	
	type_of_ruv <-  "RUVs"
	
	organism <- "scer"
		
}
	

if( !(type_of_ruv %in% c( "RUVg", "RUVs", "RUVr"))) {
	stop( "Type of RUV chosen is not valid.")
}

```

## Global parameters
```{r}


```

## Directory Management
```{r}

type_of_hypothesis_test <- ifelse(  is_quasi_likelihood_test, "QL", "LRT")


 num_samples_dir <- "All_Samples"

if( "5a" %in% samples_to_remove) {
  
  num_samples_dir <- "Remove_5a"
} 

choose_k_results_dir <- file.path(results_dir, "Differential_Expression",  
										paste(  "EdgeR",  type_of_ruv,  type_of_hypothesis_test, sep="_" ),  
										organism , 
										num_samples_dir, 
										paste("k_", num_unwanted_variables, sep="") ) 

create_dir_if_not_exists( choose_k_results_dir ) 

## Save the results into files 
edger_ruvs_gene_lists_dir <- file.path( choose_k_results_dir, "Gene_Lists")
create_dir_if_not_exists( edger_ruvs_gene_lists_dir)


```

## Load DGEList object and the normalised read counts for gene 
```{r}
scer_dgelist <- readRDS( file=file.path( choose_k_results_dir, "scer_deg_ruv.Rdata") ) 

```

## Load Gene Annotation
```{r}
gene_annotation <- read.table ( file.path( data_dir, 'Gene_Annotation/scer_gene_annotation_summary.txt' ) , header=TRUE, sep="\t", quote="", stringsAsFactors = FALSE)

```


## Load design matrix
```{r}

ruvs_design <- readRDS( file=file.path( choose_k_results_dir, "ruvs_design_matrix.Rdata") ) 

design_matrix <- read.table(scer_design_matrix_file, header=TRUE, sep="\t", stringsAsFactors = TRUE)


design_matrix <- design_matrix %>%
				 dplyr::mutate( Biological_Replicate = Replicate )

design_matrix <- design_matrix %>%
				 dplyr::mutate( Short_Name = paste(Experiment , Biological_Replicate, sep="") )

## Remove sample 5a
design_matrix <- design_matrix %>%
                  filter( !(Sample %in% samples_to_remove) )

## Fix up ordering of design matrix and sample_counts
rownames(design_matrix) <- design_matrix$Short_Name
design_matrix <- design_matrix[colnames(scer_dgelist$E),]
colnames(scer_dgelist$E) ==rownames(design_matrix)

```



## Defining each treatment combination as a group
## See page 34 and 35 of EdgeR user guide
```{r}

unwanted_factors_vector <- purrr::map_chr ( 1:num_unwanted_variables,  function(x) { paste( "W_", x, sep="") } )

colnames( ruvs_design) <- c(
  levels(design_matrix$Experiment), unwanted_factors_vector ) 

equation_string <- paste( "~ 0 + A + CA + AL + CAL + " , 
						  paste( unwanted_factors_vector, collapse=" + "), sep=" + ") 

 design_by_group <- model.matrix(  formula(uq(equation_string )), data=ruvs_design )

 
fit_by_group <- NULL
if ( is_quasi_likelihood_test == TRUE) {
	
	fit_by_group <- glmQLFit(scer_dgelist, design_by_group)

} else {
	
	fit_by_group <- glmFit(scer_dgelist, design_by_group)

}
```

## Then we can make any comparisons we wish. For example, we might wish to make the following contrasts:
```{r} 
 my.contrasts <- makeContrasts( compare_A   = A - CA,
								compare_AL   = AL - CAL,
								levels=design_by_group  )
```

## Do calculation and get list of significantly differentially expressed genes
```{r}
get_top_tags_baseed_on_group <- function ( group,  fit_by_group ) {
 	
	if ( is_quasi_likelihood_test == TRUE ) {
 		lrt <- glmQLFTest(fit_by_group, contrast=my.contrasts[,group]) 
	} else   {
 		lrt <- glmLRT(fit_by_group, contrast=my.contrasts[,group]) 
	}
 	
	top <- topTags(lrt, n=nrow(fit_by_group), sort.by="PValue")$table
	
	top <- rownames_to_column(top)
	
	print(colnames( top))
    
	if ( is_quasi_likelihood_test == TRUE ) {
	    colnames( top) <-  c("genes", "logFC", "logCPM", "F", "PValue", "FDR")
	} else   {
	    colnames( top) <-  c("genes", "logFC", "logCPM", "LR", "PValue", "FDR")
	}
	
	is.de <- decideTestsDGE(lrt)

	plotMD(
	lrt,
	status = is.de,
	values = c(1, -1),
	col = c("red", "blue"),
	legend = "topright",
	main = group,
	cex.main = 1)
	
		### plot log-fold change against log-counts per million, with DE genes highlighted
	  pdf(file.path(edger_ruvs_gene_lists_dir, paste( group, "_", "plotSmear.pdf", sep="")))
    plotSmear(lrt, de.tags=top %>% filter( FDR < de_genes_fdr_cutoff) %>% pull(genes), main=group)
    abline(h=c(-1, 1), col="blue")
    dev.off()
	
	
    ### plot log-fold change against log-counts per million, with DE genes highlighted
	  png(file.path(edger_ruvs_gene_lists_dir, paste( group, "_", "plotSmear.png", sep="")))
    plotSmear(lrt, de.tags=top %>% filter( FDR < de_genes_fdr_cutoff) %>% pull(genes), main=group)
    abline(h=c(-1, 1), col="blue")
    dev.off()
	
	return (top)
}

my_get_top_tag <- partial( get_top_tags_baseed_on_group, fit_by_group=fit_by_group)
list_of_top_tags <- purrr::map(  colnames( my.contrasts) ,  my_get_top_tag  )
names( list_of_top_tags) <- colnames( my.contrasts)
```

## Plot p-values distribution
```{r}
plot_p_values_dist <- function(input_table, plot_title ) {
  
  return( input_table %>%
  ggplot ( aes(PValue )) +
  geom_histogram()  +
  ggtitle( plot_title))
}

list_of_p_val_dist <-  purrr::map2(list_of_top_tags, names(list_of_top_tags), plot_p_values_dist )


pdf(file.path(edger_ruvs_gene_lists_dir, "p_values_distribution.pdf" ))
invisible(lapply(list_of_p_val_dist, print))
dev.off()

## Merge with gene annotation
merge_with_gene_annotation <- function ( input_table, gene_annotation) {
	
	return_table <- input_table %>%
					dplyr::left_join( gene_annotation, by=c("genes" = "oln_id")) %>%
					dplyr::distinct()
	
	return_table
}

my_merge_with_gene_annotation <- partial (merge_with_gene_annotation,  gene_annotation=gene_annotation)

list_of_top_tags_with_annot <- purrr::map ( list_of_top_tags,  my_merge_with_gene_annotation)


file_names_to_use <- purrr::map( names( list_of_top_tags_with_annot)  , function(x) {  file.path(edger_ruvs_gene_lists_dir, paste(x, ".tab" , sep="") )  } )

walk2 ( list_of_top_tags_with_annot, file_names_to_use, ~write.table( .x, file=.y , sep="\t", row.names = FALSE, quote = FALSE, na=""   ))

saveRDS(list_of_top_tags_with_annot, file=file.path(edger_ruvs_gene_lists_dir, "list_of_top_tags_table.Rdata" ))

# list_of_top_tags_with_annot <- readRDS( file=file.path(edger_ruvs_gene_lists_dir, "list_of_top_tags_table.Rdata" ))
```

## Calcuate number of significantly differentially expressed genes 
```{r}
num_sig_genes_list <- purrr::map ( list_of_top_tags_with_annot, num_significant_genes_edgeR)

num_sig_genes_list

# filter( list_of_top_tags_with_annot[["FLBvsCTRL.12h"]], FDR <0.05)
# filter( list_of_top_tags_with_annot[["FLBvsCTRL.8h"]], FDR <0.05)
# filter( list_of_top_tags_with_annot[["FLBvsCTRL.12vs8"]], FDR <0.05)

write.table(  as.data.frame( num_sig_genes_list ), file=file.path(edger_ruvs_gene_lists_dir, "num_sig_genes.tab" ), sep="\t", row.names = FALSE, quote = FALSE, na="" )

```




## Calculate number of significant genes, up and down in gene expression
```{r}

my_num_significant_genes_up_and_down <- partial( num_significant_genes_up_and_down_edgeR, fdr_threshold=de_genes_fdr_cutoff )

num_sig_genes_list_up_and_down <- purrr::map ( list_of_top_tags_with_annot, my_num_significant_genes_up_and_down)

num_sig_genes_up_and_down <- data.frame( edgeR_ruvs=unlist( num_sig_genes_list_up_and_down ) )

num_sig_genes_up_and_down <- tibble::rownames_to_column(num_sig_genes_up_and_down)

write.table(  num_sig_genes_up_and_down, file=file.path(edger_ruvs_gene_lists_dir, "num_sig_genes_up_and_down.tab" ), 
			  sep="\t", row.names = FALSE, quote = FALSE, na="")

```

## Separate list of genes that have statistically significantly up and down changes in gene expression
```{r}

split_into_up_and_down <- function ( top_tags, table_name, fdr_cutoff=0.05, gene_names_column = "genes") {
	
	significant_genes <-  dplyr::filter( top_tags , FDR < fdr_cutoff ) 
	
	up_genes <- dplyr::filter ( significant_genes, logFC > 0) %>%
				dplyr::select ( one_of(c( gene_names_column)) ) %>%
				t() %>%
				as.vector()
	
	down_genes <- dplyr::filter ( significant_genes, logFC < 0) %>%
				dplyr::select ( one_of(c( gene_names_column)) ) %>%
				t() %>%
				as.vector()
	
	return_list <- list ( up_genes, down_genes)
	
	names(return_list) <-  c( paste(table_name, "up", sep="." ) , 
							  paste(table_name, "down", sep="." ) ) 
	
	return( return_list)
	
}

# test_list <- split_into_up_and_down( list_of_top_tags_with_annot[[1]], names( list_of_top_tags_with_annot)[[1]])


my_split_into_up_and_down <- partial( split_into_up_and_down, fdr_cutoff=de_genes_fdr_cutoff, gene_names_column = "genes" )

list_of_up_and_down_genes <- map2(  list_of_top_tags_with_annot, names( list_of_top_tags_with_annot), my_split_into_up_and_down) 
list_of_up_and_down_genes <- unlist( unname(list_of_up_and_down_genes), recursive=FALSE)


saveRDS( list_of_up_and_down_genes, file=file.path(edger_ruvs_gene_lists_dir, "edgeR_up_and_down_significant_DE_genes.Rdata" ))

purrr::walk2( list_of_up_and_down_genes, names(list_of_up_and_down_genes),    
              function(x, y ) {  write.table( data.frame(oln_id= x),    
                                              file= file.path(edger_ruvs_gene_lists_dir, paste(y, ".tab", sep="" )),
                                              sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE        )         }    )


```


## Combine statistical values with experimental labels. 

For example, this table will merged with list of genes associated with enriched GO terms so that we can see the statistics and values side by side. 
```{r}

experiment_label_and_logFC_FDR_table <-  plyr::ldply( list_of_top_tags, .id="expt_label")

experiment_label_and_logFC_FDR_table <- experiment_label_and_logFC_FDR_table %>% 
										dplyr::mutate( query_list_id = paste( expt_label, 
																			  ifelse(logFC > 0 , "up", "down"), 
																			  sep="." ) ) %>%
										dplyr::distinct()

write.table ( experiment_label_and_logFC_FDR_table, 
			  file=file.path(edger_ruvs_gene_lists_dir,
				  			 "experiment_label_and_logFC_FDR_table.tab" ), sep="\t", row.names = FALSE, quote = FALSE, na="" )


FDR_table_all_tests_with_annotation <- merge_with_gene_annotation ( experiment_label_and_logFC_FDR_table, gene_annotation)

head( FDR_table_all_tests_with_annotation)


write.table ( FDR_table_all_tests_with_annotation, 
			  file=file.path(edger_ruvs_gene_lists_dir,
				  			 "FDR_table_all_tests_with_annotation.tab" ), sep="\t", row.names = FALSE, quote = FALSE, na="" )


```

## Function: draw_volcano_plot 
```{r}

draw_volcano_plot <- function ( fold_change_table, title, 
								fdr_threshold=0.05, fdr_column_name="FDR", logFC_column_name = "logFC" ) {
	
	data_frame_for_volcano_plot <- data.frame( logFC=fold_change_table[, logFC_column_name],
											   logPvalue=-log10(fold_change_table[, fdr_column_name]),
											   is_significant=ifelse ( fold_change_table[, fdr_column_name] < fdr_threshold, 
											   						TRUE, FALSE) ) 

	
	my_plot <- data_frame_for_volcano_plot %>%
			   dplyr::mutate( color_groups = ifelse(logPvalue >  -log10(0.05) & abs(logFC) <= 1, "orange", "black" )) %>%
			   dplyr::mutate( color_groups = ifelse(logPvalue <= -log10(0.05) & abs(logFC) > 1, "black", color_groups )) %>%
			   dplyr::mutate( color_groups = ifelse(logPvalue >  -log10(0.05) & abs(logFC) > 1, "red", color_groups )) %>%
			   ggplot (aes(logFC, logPvalue )) +
			   geom_point(aes(color=color_groups)) + 
			   xlab("log2 fold change") +
			   ylab("-log10 p-value") + 
			   theme(legend.position="topright") + 
			   ggtitle(title) +
			   scale_color_identity( )
			
	
	return(my_plot)
	
}


my_volcano_plot <- partial(  draw_volcano_plot, fdr_threshold = de_genes_fdr_cutoff, 
							 fdr_column_name="FDR", logFC_column_name = "logFC")

list_of_volcano_plot <- map2(list_of_top_tags, names(list_of_top_tags) , my_volcano_plot)

pdf(file.path(edger_ruvs_gene_lists_dir, "volcano_plots.pdf" ))
invisible(lapply(list_of_volcano_plot, print))
dev.off()
```




